### PR TITLE
WIP on moving sprites to the path correctly

### PR DESCRIPTION
**Pull request by darzu**
Test project:
https://makecode.com/_eCTHYFMLqg3Y

future TODO:  transition sprite gently to middle of first tile

this PR TODO:
- [x] Try dikjstra's instead of distance-squared for getting onto nearest path
- [x] Create stress test

Current state:

Notice how they move to the  path but don't follow it once they get there

### CODE REVIEW

File: path-following.ts [hunk 1]
```diff
@@ -84,6 +84,24 @@ namespace scene {
...
+        const remainingPath = nearestIdx === 0 ? path : path.filter((_, i) => i >= nearestIdx);
+        if (nearestTile.x === currentLocation.x && nearestTile.y ===  currentLocation.y) {
+            // already on the path
+            _followPath(sprite, remainingPath, speed);
+        }
+        else {
+            // off path, so path to it first
+            const pathToNearest = aStar(currentLocation, path[nearestIdx])
+            _followPath(sprite, pathToNearest, speed, () => {
+                _followPath(sprite, remainingPath, speed);
```
Comment (darzu) [inline path-following.ts:110]: For some reason this code doesn't seem to run in the callback, investigating...

Edit: fixed if I use  run-in-parallel. Would be better to have a different work around tho

Comment (darzu) [reply path-following.ts:110]: Also, notice the logic here:
When a sprite is told to follow a path,
First find the nearest path tile,
If you're on it, follow the remainder of the path,
Otherwise, run a-star to get to the path and follow that
     Once there, follow the remainder of the original path

What do you all think? @jwunderl , @richard, @shakao

Comment (darzu) [reply path-following.ts:110]: It's not optimal, but it should work for most cases and it should be pretty efficient if the sprite is already on the path or very close by

File: path-following.ts [hunk 2]
```diff
@@ -84,6 +84,24 @@ namespace scene {
...
+            })
+        }
+    }
+
+    export function teleportToAndFollowPath(sprite: Sprite, path: tiles.Location[], speed?: number) {
```
Comment (darzu) [inline path-following.ts:114]: this is the old behavior

File: path-following.ts (hunk 1 above)
Comment (jwunderl) [reply path-following.ts:110]: This will fail to find closest tile on something like this: https://makecode.com/_4RC2jiHE95hv (presuming a and b were still points on the path, it would try to go to b even though a is closest).

That doesn't matter in the given case (because they're on the same total path), but it should be easy to find the true closest tile - instead of aStar(currentLocation, path[nearestIdx]),
make it an a* with no heuristic (so just dijkstras) and make the end check be that the tile is any of the tiles in the given path. It will just move out in all directions until it reaches the any of the tiles in the list, and that's the closest reachable tile -- so

```typescript
 generalAStar(tm, start,
        () => 0,
        t => path.some(t2 => t2.x === t.x && t2.y === t.y))
```

(or so)

**Comment by darzu** [comment]
Working pretty well!

**Review by jwunderl — APPROVED** [review]
Looks good, other comment can be fixed in a separate change if you want / is a bit of an edge case

File: path-following.ts (hunk 1 above)
Comment (darzu) [reply path-following.ts:110]: Yeah I originally  had something like that, but my thought was nearest via direct distance would be good enough and more efficient when you have a lot of sprites.

I'm imagining one of the main scenarios is you're moving a path slightly that a bunch of sprites are already on. Running dijkstras for each one with a O(n) end check seems like maybe too much..  Whereas the square-distance check is just O(n) once per sprite

Comment (darzu) [reply path-following.ts:110]: If you want the  sprite to take a truly optimal path to the destination, then you should get a fresh a-star from its position anyway I think. Because even running dijkstras to get to the nearest path tile isn't as optimal as pathing yourself to the end directly.

Comment (darzu) [reply path-following.ts:110]: But maybe dijkstra's would be fast enough it doesn't matter

Comment (jwunderl) [reply path-following.ts:110]: But you still find a* to path to the tile that is nearest by square distance, and in a case where there's a new wall blocking (e.g. tower defense where the wall becomes blocked and they have to back out) it'll be significantly longer than the original search to nearest tile (in the above case probably like 3 times as long?)

Comment (darzu) [reply path-following.ts:110]: You can see the inefficient pathing you pointed out here too:

**Comment by darzu** [comment]
> Looks good, other comment can be fixed in a separate change if you want / is a bit of an edge case

I'm going to try your approach  and run a stress test and see how much we can handle

File: path-following.ts (hunk 1 above)
Comment (jwunderl) [reply path-following.ts:110]: Yeah, I would suspect it's actually significantly faster in a case like that to do the any path dijkstras than the a* to a specific one -- it does the search anyway, and in such a small area the cost between dijkstras and a* will be insignificant / you won't get the same degenerate cases where it does a search over the full map

Comment (jwunderl) [reply path-following.ts:110]: e.g. this case: https://makecode.com/_M79cXhbWtcDV

File: path-following.ts [hunk 3]
```diff
@@ -84,6 +84,26 @@ namespace scene {
     //% path.defl="locationTiles"
     //% group="Tiles" weight=9
     export function followPath(sprite: Sprite, path: tiles.Location[], speed?: number) {
+        const currentLocation = locationOfSprite(sprite)
```
Comment (jwunderl) [inline path-following.ts]: Oh, also can we add null checks on sprite & path here to gracefully handle the 'no path' case? e.g. do nothing if !path

**Comment by darzu** [comment]
Updated to use Dijkstra's to get to the path:

**Review by jwunderl — APPROVED** [review]
LGTM once it doesn't throw on `path == null` ~

**Comment by darzu** [comment]
Hmm there's definitely some buggy behavior going on that I need to look into:

**Comment by darzu** [comment]
Yeah, it's definitely buggy with the removing and add even with the run-in-parallel code.
I'll see what I can do

**Comment by darzu** [comment]
Some adversarial routing with probably a couple hundred sprites active:


Not too shabby!
A couple seconds delay on my macbook pro.
Unsure how this will perform on hardware.

**Comment by darzu** [comment]
Unsure how this will perform on hardware though

**Comment by darzu** [comment]
Alright I made a pass over the path following logic and made a few minor improvements:
1. Instead of teleporting to the start of the path, I just set the velocity to face that title
2. Instead of keeping a seperate "to remove" list, I enumerate the sprites backwards so that I can do "removeAt" and "push" without affecting the worklist. This makes it so that I don't need the "run in parallel" code anymore and also fixed some bugs around sprites not resuming pathing properly if there are lots of path switches.
3. I use vector math instead of atan2 + cos + sin to set direction; sqrt is much faster than trig functions generally
4. Fixed a bug where we weren't setting the a new "on end" callback correctly

I think things basically work well now:


There are some weird interactions when sprites are place inside of walls (e.g. placing a tower on them). Sprites don't get back onto the path. But I'm not sure the path following logic can do too much about that.

**Review by jwunderl — APPROVED** [review]
